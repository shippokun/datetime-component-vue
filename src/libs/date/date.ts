import { format, isValid, parseISO } from 'date-fns';

export const dateFormat = (
  date: Date,
  formatType = 'yyyy-MM-dd HH:mm:ss'
): string => {
  return format(date, formatType);
};

type DatetimeFormat = 'date' | 'time';

/**
 * iso8601形式で入力された日時を日付か時間を返す
 * @param datetime フォーマット対象の文字列
 * @param format 日付か時間のどちらの要素にフォーマットするか
 */
export const isoFormat = (
  datetime: string | undefined,
  format: DatetimeFormat
): string => {
  if (datetime === undefined) return '';
  if (isValid(parseISO(datetime))) {
    const date = new Date(datetime);
    if (format === 'date') return dateFormat(date, 'yyyy-MM-dd');
    else return dateFormat(date, 'HH:mm');
  } else {
    return '';
  }
};

/**
 * 日付と時間をくっつけてISO8601(UTC)形式で返す
 * @param date 日付(JST) yyyy-MM-dd
 * @param time 時間(JST) HH-mm
 */
export const joinDatetimeToISO = (date: string, time: string): string => {
  if (isValid(parseISO(date))) {
    const datetimeJST = time.match(/^\d{2}:\d{2}$/)
      ? new Date(`${date}T${time}:00+09:00`)
      : new Date(`${date}T00:00:00+09:00`);
    return datetimeJST.toISOString();
  } else {
    return '';
  }
};
