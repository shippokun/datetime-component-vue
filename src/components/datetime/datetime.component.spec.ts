import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import VuetifyType from 'vuetify/types';

import DatetimeComponent from './datetime.component.vue';
import * as DateLibs from '@/libs/date';

jest.mock('@/libs/date', () => ({
  isoFormat: jest.fn(),
  joinDatetimeToISO: jest.fn()
}));

const localVue = createLocalVue();
localVue.use(Vuetify);

describe('datetime.component.vue', () => {
  let vuetify: VuetifyType;

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('render', () => {
    const wrapper = mount(DatetimeComponent, {
      localVue,
      vuetify,
      propsData: {
        value: ''
      }
    });

    expect(wrapper.html()).toMatchSnapshot();
  });
  it('call emitted', async () => {
    const isoFormatMock = jest
      .spyOn(DateLibs, 'isoFormat')
      .mockImplementation(() => '');
    const joinDateTimeToISOMock = jest.spyOn(DateLibs, 'joinDatetimeToISO');

    const wrapper = mount(DatetimeComponent, {
      localVue,
      vuetify,
      propsData: {
        value: ''
      }
    });

    expect(isoFormatMock.mock.calls.length).toBe(2);
    expect(isoFormatMock.mock.calls[0]).toEqual(['', 'date']);
    expect(isoFormatMock.mock.calls[1]).toEqual(['', 'time']);

    expect(joinDateTimeToISOMock.mock.calls.length).toBe(1);
    expect(joinDateTimeToISOMock.mock.calls[0]).toEqual(['', '']);

    const inputData = '09:00';
    jest.spyOn(DateLibs, 'joinDatetimeToISO').mockReturnValue(inputData);
    const timeInput = wrapper.find('input[type="time"]');
    await timeInput.setValue(inputData);
    expect(
      (wrapper.find('input[type="time"]').element as HTMLInputElement).value
    ).toBe(inputData);

    expect(wrapper.emitted('input')).toBeTruthy();
    expect(wrapper.emitted('input')?.[0]).toStrictEqual([inputData]);

    const nextProp = '1990-01-01';
    await wrapper.setProps({ value: nextProp });

    expect(isoFormatMock.mock.calls.length).toBe(4);
    expect(isoFormatMock.mock.calls[2]).toEqual([nextProp, 'date']);
    expect(isoFormatMock.mock.calls[3]).toEqual([nextProp, 'time']);
  });
});
