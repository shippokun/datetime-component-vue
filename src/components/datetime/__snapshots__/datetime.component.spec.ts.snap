// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`datetime.component.vue render 1`] = `
<div class="row row--dense align-center">
  <div class="col">
    <div class="v-menu">
      <div class="v-input v-input--is-readonly theme--light v-text-field">
        <div class="v-input__prepend-outer">
          <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
        </div>
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><input role="button" aria-haspopup="true" aria-expanded="false" id="input-4" readonly="readonly" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
      <!---->
    </div>
  </div>
  <div class="col">
    <div class="v-input v-input--is-label-active theme--light v-text-field">
      <div class="v-input__control">
        <div class="v-input__slot">
          <div class="v-text-field__slot"><input id="input-8" type="time"></div>
        </div>
        <div class="v-text-field__details">
          <div class="v-messages theme--light">
            <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
